server/form-pages/apply/risk-and-need-factors/location-factors/preferredAps.ts

```diff
@@ -0,0 +1,99 @@
+import type { DataServices, SelectOption, TaskListErrors } from '@approved-premises/ui'
+import { ApprovedPremisesApplication as Application } from '../../../../@types/shared'
+
+import { Page } from '../../../utils/decorators'
+
+import TasklistPage from '../../../tasklistPage'
+import { numberToOrdinal } from '../../../../utils/utils'
+
+export type PreferredApsBody = {
+  preferredAp1: string
+  preferredAp2: string
+  preferredAp3: string
+  preferredAp4: string
+  preferredAp5: string
+  selectedAps: Array<SelectOption>
+}
+
+const preferredAps = new Array(5).fill('').map((_, i) => `preferredAp${i + 1}`)
+
+@Page({
+  name: 'preferred-aps',
+  bodyProperties: [...preferredAps, 'selectedAps'],
+})
+export default class PreferredAps implements TasklistPage {
+  title = 'Select a preferred AP'
+
+  premisesSelectList: Array<SelectOption> = []
+
+  preferredApOptions = preferredAps
+
+  preferredApLabels = this.preferredApOptions.map((_, i) => `${numberToOrdinal(i)} choice AP`)
+
+  selectedAps: Array<SelectOption>
+
+  body: PreferredApsBody
+
+  constructor(private _: Partial<PreferredApsBody>) {}
+
+  static async initialize(
+    body: Record<string, unknown>,
+    _: Application,
+    token: string,
+    { premisesService }: DataServices,
+  ) {
+    const premisesSelectList = [
+      { text: 'No preference', value: 'no-preference' },
+      ...(await premisesService.getPremisesSelectList(token)),
```

Comment:
I find using `getPremisesSelectList` a bit confusing here, as we're mixing presentation logic in with other logic. I'd feel happier if we just added a `getAll` method to the PremisesService and then used that. Then we could use `convertObjectsToSelectOptions` on the list of premises in the view layer.

Reply:
yeah fair enough ill make that change